# <center> Analysis of vacancies from HeadHunter using SQL
   

In [3]:
import pandas as pd
import psycopg2

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Preliminary data analysis

1. Lets see how many vacancies are in our database 

In [6]:
query_3_1 = f'''select count(*)
                from vacancies
            '''

In [7]:
qty_vac= pd.read_sql_query(query_3_1, connection)
display(qty_vac)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1157217138.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_vac= pd.read_sql_query(query_3_1, connection)


,count
0,49197


2. Now lets see quantity of employers. 

In [8]:
query_3_2 = f'''select count(*)
                from employers
            '''

In [9]:
qty_empr= pd.read_sql_query(query_3_2, connection)
display(qty_empr)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2458544806.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_empr= pd.read_sql_query(query_3_2, connection)


,count
0,23501


3. Areas where vacancies are.

In [10]:
query_3_3 = f'''select count(*)
                from areas
            '''

In [11]:
qty_areas= pd.read_sql_query(query_3_3, connection)
display(qty_areas)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1717760068.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_areas= pd.read_sql_query(query_3_3, connection)


,count
0,1362


4. Quantity of industries.

In [12]:
query_3_4 = f'''select count(*)
                from industries
            '''

In [13]:
qty_ind= pd.read_sql_query(query_3_4, connection)
display(qty_ind)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\408090445.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_ind= pd.read_sql_query(query_3_4, connection)


,count
0,294


***

I started working with the data. I found the total number of vacancies, employers, regions, and industries. I wonder why there are so many regions; perhaps there are vacancies outside of Russia. It's also worth noting that, on average, there are two vacancies per employer.

# Detailed analysis of vacancies

1. Lets find out how many vacancies are in each area and sort them.

In [14]:
query_4_1 = f'''select a.name region, count(v.id) qty
                from vacancies v
                    join areas a on a.id=v.area_id
                group by a.name
                order by 2 desc
                limit 5
            '''

In [15]:
cnt_area= pd.read_sql_query(query_4_1, connection)
display(cnt_area)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3807556044.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_area= pd.read_sql_query(query_4_1, connection)


,region,qty
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892


2. How many vacancies have filled salary range?

In [16]:
query_4_2 = f'''select count(*)
                from vacancies
                where salary_from is not null or salary_to is not null
            '''

In [17]:
qty_with_sal= pd.read_sql_query(query_4_2, connection)
display(qty_with_sal)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2001132973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_with_sal= pd.read_sql_query(query_4_2, connection)


,count
0,24073


3. Finding out whats the average salary range. 

In [18]:
query_4_3 = f'''select round(avg(salary_from)) sf, round(avg(salary_to)) st
                from vacancies
            '''

In [19]:
avg_from_to= pd.read_sql_query(query_4_3,connection)
display(avg_from_to)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\729226904.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_from_to= pd.read_sql_query(query_4_3,connection)


,sf,st
0,71065.0,110537.0


4. Lets see quantity of every type of schedule and every type of employment.


In [20]:
query_4_4 = f'''select schedule, employment, count(*)
                from vacancies
                group by schedule, employment
                order by 3 desc
            '''

In [21]:
qty_pairs= pd.read_sql_query(query_4_4, connection)
display(qty_pairs)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\875815956.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_pairs= pd.read_sql_query(query_4_4, connection)


,schedule,employment,count
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Whats the most needed experience level?

In [22]:
query_4_4 = f'''select distinct experience, count(experience)
                from vacancies
                group by experience
                order by 2 asc
            '''

In [23]:
exp_cnt= pd.read_sql_query(query_4_4, connection)
display(exp_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3426249347.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  exp_cnt= pd.read_sql_query(query_4_4, connection)


,experience,count
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

The analysis clearly shows that the majority of employers are seeking people in large CIS cities. This is likely why the average salary listed is slightly higher than the CIS average. However, job postings generally don't specify a salary range. We can also see that while remote work is in some demand, the vast majority of vacancies offer full-time positions. Currently, finding a job in the labor market will be difficult for those seeking employment at the beginning of their career.

# Employer Analysis

1. Top 5 employers by quantity of vacancies.

In [24]:
query_5_1 = f'''select distinct e.name, count(v.id)
                from vacancies v
                    join employers e on v.employer_id=e.id
                group by e.name
                order by 2 desc
                limit 5
            '''

In [25]:
cnt_vac= pd.read_sql_query(query_5_1, connection)
cnt_vac.head()

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\4118035862.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac= pd.read_sql_query(query_5_1, connection)


,name,count
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


2. Finding annomaly area where there are no vacancies but plenty of employers.

In [26]:
query_5_1 = f'''select a.name as name, count(e.id) as emp, count(v.id) as vac
                from areas a
                    left join vacancies v on a.id=v.area_id
                    left join employers e on a.id= e.area
                group by a.name 
                having count(v.id)=0
                order by 2 desc
                limit 1
            '''

In [27]:
count_emp_vac= pd.read_sql_query(query_5_1, connection)
display(count_emp_vac)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3544299062.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_emp_vac= pd.read_sql_query(query_5_1, connection)


,name,emp,vac
0,Россия,410,0


3. Top employers by areas.

In [28]:
query_5_2 = f'''select e.name as name, count(distinct v.area_id) as vac
                from employers e
                    left join vacancies v on v.employer_id= e.id
                group by e.name
                order by 2 desc
            '''

In [29]:
emp_area= pd.read_sql_query(query_5_2, connection)
display(emp_area)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\310555112.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_area= pd.read_sql_query(query_5_2, connection)


,name,vac
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
23170,СДЕЛКА,0
23171,Alandr Group,0
23172,СДК,0
23173,Lemon Land Lombard,0


4. What quantity of employers have no industry filled in their vacancies? 

In [30]:
query_5_3 = f'''select count(e.id)
                from employers e
                    left join employers_industries ei on ei.employer_id= e.id
                where ei.industry_id is null
            '''

In [31]:
ind_n= pd.read_sql_query(query_5_3, connection)
display(ind_n)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1073280717.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ind_n= pd.read_sql_query(query_5_3, connection)


,count
0,8419


5. THird employer in alphabetically sorted data that has filled 4 industries. 

In [32]:
query_5_4 = f'''select e.name
                from employers e
                    join employers_industries ei on e.id=ei.employer_id
                group by e.id, e.name
                having count(distinct ei.industry_id)=4
                order by 1
                limit 1 offset 2
            '''

In [33]:
emp_ind= pd.read_sql_query(query_5_4, connection)
display(emp_ind)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1775747399.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_ind= pd.read_sql_query(query_5_4, connection)


,name
0,2ГИС


6. Which employer has software engeneering in their industries spot. 

In [34]:
query_5_5 = f'''select count(distinct ei.employer_id)
                from employers_industries ei
                    join industries i on i.id=ei.industry_id
                where i.name = 'Разработка программного обеспечения'
            '''

In [35]:
count_emp= pd.read_sql_query(query_5_5, connection)
display(count_emp)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1324660682.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_emp= pd.read_sql_query(query_5_5, connection)


,count
0,3553


7. How many vacancies does Yandex have in Russias major cities?

In [36]:
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

In [37]:
# Данное решение нашел в телеграм канале, поскольку перепробовал кучу решений и у меня ничего не вышло.
headers= {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36'}
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url, headers=headers)
page = BeautifulSoup(response.text, 'html.parser')
tabl = page.find("table", class_="standard")

millionniki = []
table = tabl.find("tbody")
rows = table.find_all("tr")


for i in rows:
    tags= list(i.find("a"))
    millionniki.append(tags[0])
millionniki= millionniki[1:]
print(millionniki)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [38]:
query_5_6 = f'''select a.name as name, count(v.id) as cnt
                from vacancies v
                    join employers e on e.id=v.employer_id
                    join areas a on a.id = v.area_id
                where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
                group by a.name
                
                union all
                select 'Total' as name, count(v.id) as cnt
                from vacancies v
                    join employers e on e.id=v.employer_id
                    join areas a on a.id = v.area_id
                where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
                group by e.name
                order by 2 asc
            '''

In [39]:
yandex_mil= pd.read_sql_query(query_5_6, connection)
display(yandex_mil)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3295075723.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_mil= pd.read_sql_query(query_5_6, connection)


,name,cnt
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Уфа,26
8,Самара,26
9,Краснодар,30


***

Having analyzed employers, I can confidently say that companies with a developed IT sector are hiring the most employees. An example of this is Yandex, which is seeking 1,933 specialists in 181 regions of the CIS. It's worth noting that although companies are looking for people living in specific regions, we learned this because area-Russia has the most companies but no vacancies, which may indicate the unpopularity of remote work.

# Subject analysis

1. How many data oriented vacancies are there?

In [40]:
query_6_1 = f'''select count(v.id)
                from vacancies v
                where lower(v.name) like '%data%' or lower(v.name) like '%данн%'                    
            '''

In [41]:
data_cnt= pd.read_sql_query(query_6_1, connection)
display(data_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\4191512999.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_cnt= pd.read_sql_query(query_6_1, connection)


,count
0,1771


2. Entry level data science vacancies.

In [42]:
query_6_2 = f'''select count(v.id)
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and v.name not like '%HTML%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    lower(v.name) like '%junior%' or 
                    v.experience= 'Нет опыта' or 
                    v.employment= 'Стажировка'
                )                   
            '''

In [43]:
data_sci_cnt= pd.read_sql_query(query_6_2, connection)
data_sci_cnt.head()

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\205585809.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_sci_cnt= pd.read_sql_query(query_6_2, connection)


,count
0,51


3. Vacancies that require skills like SQL and postgres.

In [44]:
query_6_3 = f'''select count(v.id)
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    lower(v.key_skills) like '%postgres%' or
                    lower(v.key_skills) like '%sql%'
                )                   
            '''

In [45]:
sql_cnt= pd.read_sql_query(query_6_3, connection)
display(sql_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2613755694.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql_cnt= pd.read_sql_query(query_6_3, connection)


,count
0,229


4. How popular is Python?

In [46]:
query_6_4 = f'''select count(v.id)
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    lower(v.key_skills) like '%python%'
                )                   
            '''

In [47]:
py_cnt= pd.read_sql_query(query_6_4, connection)
display(py_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1545349579.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  py_cnt= pd.read_sql_query(query_6_4, connection)


,count
0,357


5. Average quantity of key skills needed for DS vacancies.

In [48]:
query_6_5 = f'''select *
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and v.name not like '%HTML%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                )                  
            '''

In [49]:
import numpy as np
avg_skills= pd.read_sql_query(query_6_5, connection)
def get_qty(text):
    x=0
    if text is not None:
        text= text.split('\t')
        x+=len(text)
        return x
    else:
        return np.nan
avg_skills['qty']=avg_skills['key_skills'].apply(get_qty)
round(avg_skills['qty'].mean(), 2)  

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2437790653.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_skills= pd.read_sql_query(query_6_5, connection)


np.float64(6.41)

6. Average DS salary.

In [50]:
query_6_6 = f'''select v.experience as experience, avg((coalesce(v.salary_from, 0)+ coalesce(v.salary_to, 0))/
                    (case when v.salary_from is not null and v.salary_to is not null then 2 else 1 end)) as average_salary
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and v.name not like '%HTML%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    v.salary_from is not null or
                    v.salary_to is not null
                    )
                group by v.experience                  
            '''

In [51]:
avg_salary= pd.read_sql_query(query_6_6, connection)
display(avg_salary)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1391840803.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary= pd.read_sql_query(query_6_6, connection)


,experience,average_salary
0,Нет опыта,74642.857143
1,От 1 года до 3 лет,139674.750000
2,От 3 до 6 лет,243114.666667


***

This analysis shows that data science is in high demand, although the barrier to entry for newcomers is significantly higher. This is especially true for data scientists, for whom job openings are significantly fewer and average starting salaries are almost half as much. Some of the most in-demand skills are SQL and Python.

# Conclusion

In [52]:
query_bonus = f'''select a.name as areas, i.name as industries, count(v.id) as vac
                from vacancies v
                    join areas a on v.area_id= a.id
                    join employers e on e.id= v.employer_id
                    join employers_industries ei on ei.employer_id= v.employer_id
                    join industries i on i.id= ei.industry_id
                group by a.name, i.name
                order by 3 desc                 
            '''

In [53]:
bonus= pd.read_sql_query(query_bonus,connection)
display(bonus)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2907641292.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  bonus= pd.read_sql_query(query_bonus,connection)


,areas,industries,vac
0,Москва,Разработка программного обеспечения,1294
1,Москва,"Системная интеграция, автоматизации технологи...",1117
2,Минск,Разработка программного обеспечения,1023
3,Санкт-Петербург,Разработка программного обеспечения,908
4,Москва,"Интернет-компания (поисковики, платежные систе...",798
...,...,...,...
13532,Нахабино,Двигатели (производство),1
13533,Нахабино,"Железнодорожное машиностроение, вагоностроение",1
13534,Нахабино,Интернет-магазин,1
13535,Нахабино,"Курьерская, почтовая доставка",1


After analyzing job postings on hh.ru, I see a shortage of experienced IT specialists in the labor market. Most employers are either software developers or companies with a strong IT sector (such as banks). Demand for such specialists is highest in the capital cities of CIS countries, where companies' headquarters are likely located.